autotest/cpp/test_ogr.cpp

```diff
@@ -2163,4 +2164,82 @@ namespace
         EXPECT_EQ(i, oFDefn.GetGeomFieldCount());
     }
 
+    // Test GDALDataset QueryLoggerFunc callback
+    TEST_F(test_ogr, GDALDatasetSetQueryLoggerFunc)
+    {
+      if( GDALGetDriverByName("GPKG") == nullptr )
+      {
+          GTEST_SKIP() << "GPKG driver missing";
+      }
+
+      GDALDatasetH ds { GDALOpenEx( "data/poly-1-feature.gpkg", GDAL_OF_VECTOR|GDAL_OF_UPDATE, nullptr, nullptr, nullptr )};
+      ASSERT_TRUE( ds );
+
+      struct QueryLogEntry
+      {
+          std::string sql;
+          std::string error;
+          int64_t numRecords;
+          int64_t executionTimeMilliseconds;
+      };
+
+      std::vector<QueryLogEntry> queryLog;
+      const bool retVal = GDALDatasetSetQueryLoggerFunc( ds , [ ] (const char *pszSQL, const char *pszError, int64_t lNumRecords, int64_t lExecutionTimeMilliseconds, void *pQueryLoggerArg)
+      {
+        std::vector<QueryLogEntry>* queryLogLocal { reinterpret_cast<std::vector<QueryLogEntry>*>( pQueryLoggerArg ) };
+        QueryLogEntry entryLocal;
+        entryLocal.sql = pszSQL;
+        entryLocal.numRecords = lNumRecords;
+        entryLocal.executionTimeMilliseconds = lExecutionTimeMilliseconds;
+        if ( pszError )
+        {
+          entryLocal.error = pszError;
+        }
+        queryLogLocal->push_back( entryLocal );
+      }, &queryLog );
+
+      ASSERT_TRUE( retVal );
+      auto hLayer { GDALDatasetGetLayer( ds, 0 ) };
+      ASSERT_TRUE( hLayer );
+      ASSERT_EQ( strcmp( OGR_L_GetName( hLayer ), "poly" ), 0 );
+      OGRFeatureH hFeature { OGR_L_GetNextFeature( hLayer ) };
```

Comment:
it will leak. You might want to do something like:
```suggestion
      auto poFeature = std::unique_ptr<OGRFeature>(OGRFeature::FromHandle(OGR_L_GetNextFeature( hLayer )));
      auto hFeature = OGRFeature::ToHandle(poFeature.get());
```

Reply:
done